src/main/java/net/snowflake/client/core/auth/oauth/OAuthClientCredentialsAccessTokenProvider.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2024 Snowflake Computing Inc. All rights reserved.
+ */
+
+package net.snowflake.client.core.auth.oauth;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.nimbusds.oauth2.sdk.ClientCredentialsGrant;
+import com.nimbusds.oauth2.sdk.Scope;
+import com.nimbusds.oauth2.sdk.TokenRequest;
+import com.nimbusds.oauth2.sdk.auth.ClientAuthentication;
+import com.nimbusds.oauth2.sdk.auth.ClientSecretBasic;
+import com.nimbusds.oauth2.sdk.auth.Secret;
+import com.nimbusds.oauth2.sdk.id.ClientID;
+import java.net.URI;
+import net.snowflake.client.core.HttpUtil;
+import net.snowflake.client.core.SFException;
+import net.snowflake.client.core.SFLoginInput;
+import net.snowflake.client.core.SnowflakeJdbcInternalApi;
+import net.snowflake.client.jdbc.ErrorCode;
+import net.snowflake.client.log.SFLogger;
+import net.snowflake.client.log.SFLoggerFactory;
+
+@SnowflakeJdbcInternalApi
+public class OAuthClientCredentialsAccessTokenProvider implements AccessTokenProvider {
+
+  private static final SFLogger logger =
+      SFLoggerFactory.getLogger(OAuthClientCredentialsAccessTokenProvider.class);
+
+  private final ObjectMapper objectMapper = new ObjectMapper();
+
+  @Override
+  public String getAccessToken(SFLoginInput loginInput) throws SFException {
+    try {
+      logger.debug("Starting OAuth authorization code authentication flow...");
+      TokenRequest tokenRequest = buildTokenRequest(loginInput);
+      TokenResponseDTO tokenResponse = requestForAccessToken(loginInput, tokenRequest);
+      return tokenResponse.getAccessToken();
+    } catch (Exception e) {
+      logger.error(
+          "Error during OAuth client credentials code flow. Verify configuration passed to driver and IdP (URLs, grant types, scope, etc.)",
+          e);
+      throw new SFException(e, ErrorCode.OAUTH_CLIENT_CREDENTIALS_FLOW_ERROR, e.getMessage());
+    }
+  }
+
+  private TokenResponseDTO requestForAccessToken(SFLoginInput loginInput, TokenRequest tokenRequest)
+      throws Exception {
+    URI requestUri = tokenRequest.getEndpointURI();
+    logger.debug(
+        "Requesting OAuth access token from: {}", requestUri.getAuthority() + requestUri.getPath());
+    String tokenResponse =
+        HttpUtil.executeGeneralRequest(
+            OAuthUtil.convertToBaseRequest(tokenRequest.toHTTPRequest()),
+            loginInput.getLoginTimeout(),
+            loginInput.getAuthTimeout(),
+            loginInput.getSocketTimeoutInMillis(),
+            0,
```

Comment:
what's the 0 here?

Reply:
It's retry count. In case of SAML it's also 0 so I wanted this to be consistent.